deliverable 4

mayralgr opened the pull request:
[preview](https://fanciful-gnome-7b543f.netlify.app/)

- [x] Product Detail Page (with “Add to Cart” functionality)

- [x] Shopping Cart Page

- [x] Checkout Page -Related topics: React Router, Global State Management, Design Patterns, Best Practices & Performance Optimizations



- [x] If you don’t have it already, add a **Shopping Cart **Icon to your header. 
  - [x] badge with the number of items that you have added to your cart. 
  - [x] This icon will also serve as a link to navigate to the Shopping Cart Page after clicking on it.
  
  

- [x] Product Detail Page
  - [x] Add to Cart functionality for the selected product
  - [x] should be able to add multiple items to your cart using the quantity selector. Please don’t forget to validate that you don’t exceed the stock units available for the selected product.
  - [x] After clicking on the Add to Cart button the items should be added to your cart and the badge on the “Shopping Cart Icon” in the header should be updated to display the current quantity of items in the cart.
  - [x] The Add to Cart button should be hidden or disabled when the **stock** units available for the selected product is zero. So you shouldn’t be able to add to cart a product that is not available in the stock.

- [x]  Shopping Cart Page
  - [x]  route for this page should be **/cart**
  - [x]  order summary table
    - [x]  row in this table per item/product in the cart.
    - [x]  Each row should show the main image of the product, its name, unit price, a quantity selector, subtotal (unit price x quantity) and a Remove from cart icon.
    - [x] 3.2.3. At the bottom of the table there should be a label to display the cart total (sum of the subtotal’s column in the table) and a Proceed to checkout button that will serve as a link to navigate to the Checkout Page after clicking on it.
    - [x] 3.2.4. You should be able to modify the quantity of items that you want using the quantity selector. Please don’t forget to validate that you don’t exceed the stock units available for the selected product.
    - [x] 3.2.5. After updating the quantity the subtotal for the product and the cart total labels should be updated.


- [x] Checkout Page
  - [x]  The route for this page should be **/checkout** 
  - [x] Form to capture the customer information for their order 
    - [x]  name of the customer
    - [x]  email of the customer
    - [x] post/zip code of the customer
    - [x]  order notes
  - [x] 4.2.2.3. At the bottom of the table there should be a label to display the cart total (sum of the subtotal’s column) and two buttons, one for Place order and the other for Go back to cart
  - [x] don’t have to implement the logic for the Place order button yet.
  - [x]  Go back to cart button, you should be able to navigate to the Shopping Cart Page
  
Drag
- [x] fetch more products in the product list page in order to be able to filter by al the possible categories because some are no showing. 
- [x] global context to deal with the cart items for the final deliverable.


Bonus
The bonus is a plus that helps us know you better and increments your final score. Consider the following:

- [x] PropTypes are implemented correctly for all your components (10 points)
Error Boundaries are applied correctly at least for 1 component within the app (15 points)
useMemo and useCallback hooks are implemented correctly to improve performance if needed (20 points)

gerardo-santos reviewed (approved):
Very good job! Just two comments.

1. Maybe it would be cool to add a "cursor: pointer" to the trashcan icon when you delete a product from the cart.

2. I believe that there is a bug when you add a product to the cart, and then you delete that product from the cart and then you want to add that product to the cart again

mayralgr commented:
@gerardo-santos I added the cursor, I am not able to reproduced the bug but I think it may be due to the area of the icon, I moved some things around. If it is still happening can you let me know please

gerardo-santos commented:
> @gerardo-santos I added the cursor, I am not able to reproduced the bug but I think it may be due to the area of the icon, I moved some things around. If it is still happening can you let me know please

Don't worry, I checked it out today and it worked fine. Nice job!
